hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metrics/HoodieMetrics.java

```diff
@@ -383,27 +396,67 @@ public void updateClusteringFileCreationMetrics(long durationInMs) {
+
+    updateLatestCompletedInstant(activeTimeline, LATEST_COMPLETED_CLUSTERING_INSTANT_STR, HoodieTimeline.REPLACE_COMMIT_ACTION);
+    updateLatestCompletedInstant(activeTimeline, LATEST_COMPLETED_COMPACTION_INSTANT_STR, HoodieTimeline.COMMIT_ACTION);
+    updateLatestCompletedInstant(activeTimeline, LATEST_COMPLETED_CLEAN_INSTANT_STR, HoodieTimeline.CLEAN_ACTION);
+    updateLatestCompletedInstant(activeTimeline, LATEST_COMPLETED_ROLLBACK_INSTANT_STR, HoodieTimeline.ROLLBACK_ACTION);
+
+    updatePendingInstantCount(activeTimeline, PENDING_CLUSTERING_INSTANT_COUNT_STR, HoodieTimeline.CLUSTERING_ACTION);
+    updatePendingInstantCount(activeTimeline, PENDING_COMPACTION_INSTANT_COUNT_STR, HoodieTimeline.COMPACTION_ACTION);
+    updatePendingInstantCount(activeTimeline, PENDING_CLEAN_INSTANT_COUNT_STR, HoodieTimeline.CLEAN_ACTION);
+    updatePendingInstantCount(activeTimeline, PENDING_ROLLBACK_INSTANT_COUNT_STR, HoodieTimeline.ROLLBACK_ACTION);
+  }
+
+  private void updateEarliestPendingInstant(final HoodieActiveTimeline activeTimeline,
+                                            final String metricName,
+                                            final String action) {
+    Set<String> validActions = CollectionUtils.createSet(action);
+    HoodieTimeline filteredInstants = activeTimeline.filterInflightsAndRequested().filter(instant -> validActions.contains(instant.getAction()));
+    Option<HoodieInstant> hoodieInstantOption = filteredInstants.firstInstant();
+    if (hoodieInstantOption.isPresent()) {
+      updateMetric(action, metricName, Long.parseLong(hoodieInstantOption.get().requestedTime()));
+    }
+  }
 
+  private void updateLatestCompletedInstant(final HoodieActiveTimeline activeTimeline,
+                                            final String metricName,
+                                            String action) {
+    String tableType = config.getTableType().name();
+    if (HoodieTableType.MERGE_ON_READ.name().equalsIgnoreCase(tableType)
+        && LATEST_COMPLETED_COMPACTION_INSTANT_STR.equalsIgnoreCase(metricName)) {
+      action = HoodieActiveTimeline.COMMIT_ACTION;
+    }
+    if (HoodieTableType.COPY_ON_WRITE.name().equalsIgnoreCase(tableType)
+        && LATEST_COMPLETED_CLUSTERING_INSTANT_STR.equalsIgnoreCase(metricName)) {
+      action = HoodieActiveTimeline.REPLACE_COMMIT_ACTION;
+    }
+    Set<String> validActions = CollectionUtils.createSet(action);
+    HoodieTimeline filteredInstants = activeTimeline.filterCompletedInstants().filter(instant -> validActions.contains(instant.getAction()));
+    Option<HoodieInstant> hoodieInstantOption = filteredInstants.lastInstant();
+    if (hoodieInstantOption.isPresent()) {
+      updateMetric(action, metricName, Long.parseLong(hoodieInstantOption.get().requestedTime()));
+    }
+  }
+
+  private void updatePendingInstantCount(final HoodieActiveTimeline activeTimeline,
```

Comment:
could we add some java docs? what does `PendingInstantCount` mean? and what is this metric used for?